virtual-programs/shapes.folk

```diff
@@ -1,111 +1,301 @@
+
+When /someone/ wishes /page/ draws a set of points /points/ with /...options/ & /page/ has region /r/ {
+  set radius [dict_getdef $options radius 5]
+  set color [dict_getdef $options color white]
+  set filled [dict_getdef $options filled true]
+  set thickness [dict_getdef $options thickness 2]
+  set layer [dict_getdef $options layer 0]
+  
+  lassign [region centroid $r] cx cy
+  set angle [region angle $r]
+  set center [list $cx $cy]
+  
+  if {[dict exists $options offset]} {
+    set offset [dict get $options offset]
+    set offset [process_offset $offset $r]
+    set center [vec2 add $center [vec2 rotate $offset $angle]]
+  }
+  
+  foreach point $points {
+    set pointPos [vec2 add $center [vec2 rotate $point $angle]]
+    
+    Wish to draw a circle with center $pointPos radius $radius thickness $thickness \
+      color $color filled $filled layer $layer
+  }
+}
+
+When /someone/ wishes /page/ draws a polyline /points/ with /...options/ & /page/ has region /r/ {
+  set color [dict_getdef $options color white]
+  set thickness [dict_getdef $options thickness 2]
+  set layer [dict_getdef $options layer 0]
+  set dashed [dict_getdef $options dashed false]
+  set dashlength [dict_getdef $options dashlength 20]
+  set dashoffset [dict_getdef $options dashoffset 0]
+  
+  lassign [region centroid $r] cx cy
+  set angle [region angle $r]
+  set center [list $cx $cy]
+  
+  if {[dict exists $options offset]} {
+    set offset [dict get $options offset]
+    set offset [process_offset $offset $r]
+    set center [vec2 add $center [vec2 rotate $offset $angle]]
+  }
+  
+  set transformedPoints {}
+  foreach point $points {
+    lappend transformedPoints [vec2 add $center [vec2 rotate $point $angle]]
+  }
+  
+  if {$dashed} {
+    Wish to draw a dashed stroke with points $transformedPoints color $color width $thickness \
+      dashlength $dashlength dashoffset $dashoffset layer $layer
+  } else {
+    Wish to draw a stroke with points $transformedPoints color $color width $thickness layer $layer
+  }
+}
+
 Claim $this has demo {
+  # Center circle
   Wish $this draws a circle
-  Wish $this draws a triangle with color skyblue
-  Wish $this draws a triangle with color green offset {280 0}
-  Wish $this draws a pentagon with color gold offset {200 0}
-  Wish $this draws an octagon with color red offset {250 80}
-
+  
+  # Grid of shapes with varying thickness
+  set baseX -850
+  set baseY -200
+  set gridSpacing 130
+  
+  # Row 1: Regular polygons with different colors and thickness
+  Wish $this draws a triangle with color skyblue thickness 2 offset [list $baseX [expr {$baseY}]]
+  Wish $this draws a square with color green thickness 4 offset [list [expr {$baseX + $gridSpacing}] [expr {$baseY}]]
+  Wish $this draws a pentagon with color gold thickness 6 offset [list [expr {$baseX + $gridSpacing*2}] [expr {$baseY}]]
+  Wish $this draws a hexagon with color orange thickness 8 offset [list [expr {$baseX + $gridSpacing*3}] [expr {$baseY}]]
```

Comment:
I don't think this mode (bare coordinate numbers) makes sense, and I don't think we should support it at the shape level. It should be centimeters/meters in real space. The percentage-of-region below seems more reasonable.

Reply:
I guess that is also true of the size and thickness of shapes -- it won't make sense to have them in pixels -- so maybe we can punt on this for now, but we definitely will need to revisit it

Reply:
I agree with this, I added the percentage-of-region as a more natural mode. Will make an issue to refactor this (and all userland units) to be real-world units.